crates/egui_extras/src/loaders/gif_loader.rs

```diff
@@ -0,0 +1,135 @@
+use std::{io::Cursor, mem::size_of, sync::Arc, time::Duration};
+
+/// Array of Frames and the duration for how long each frame should be shown
+#[derive(Debug, Clone)]
+pub struct AnimatedImage {
+    frames: Vec<Arc<ColorImage>>,
+    frame_durations: GifFrameDurations,
+}
+
+impl AnimatedImage {
+    pub fn byte_len(&self) -> usize {
+        size_of::<Self>()
+            + self
+                .frames
+                .iter()
+                .map(|image| {
+                    image.pixels.len() * size_of::<egui::Color32>() + size_of::<Duration>()
+                })
+                .sum::<usize>()
+    }
+
+    /// Gets image at index
+    pub fn get_image(&self, index: usize) -> Arc<ColorImage> {
+        self.frames[index % self.frames.len()].clone()
+    }
+}
+type Entry = Result<Arc<AnimatedImage>, String>;
+
+#[derive(Default)]
+pub struct GifLoader {
+    cache: Mutex<HashMap<String, Entry>>,
+}
+
+impl GifLoader {
+    pub const ID: &'static str = egui::generate_loader_id!(GifLoader);
+}
+
+pub fn gif_to_sources(data: Bytes) -> Result<AnimatedImage, String> {
+    let decoder = image::codecs::gif::GifDecoder::new(Cursor::new(data))
+        .map_err(|_err| "Couldnt decode gif".to_owned())?;
+    let mut images = vec![];
+    let mut durations = vec![];
+    for frame in decoder.into_frames() {
+        let frame = frame.map_err(|_err| "Couldnt decode gif".to_owned())?;
+        let img = frame.buffer();
+        let pixels = img.as_flat_samples();
+
+        let delay: Duration = frame.delay().into();
+        images.push(Arc::new(ColorImage::from_rgba_unmultiplied(
+            [img.width() as usize, img.height() as usize],
+            pixels.as_slice(),
+        )));
+        durations.push(delay);
+    }
+    Ok(AnimatedImage {
+        frames: images,
+        frame_durations: GifFrameDurations(Arc::new(durations)),
+    })
+}
+
+impl ImageLoader for GifLoader {
+    fn id(&self) -> &str {
+        Self::ID
+    }
+
+    fn load(&self, ctx: &egui::Context, frame_uri: &str, _: SizeHint) -> ImageLoadResult {
+        let uri_index = decode_gif_uri(frame_uri).map_err(|e| LoadError::Loading(e.to_owned()));
+        let image_uri = uri_index.as_ref().map(|v| v.0).unwrap_or(frame_uri);
+        let mut cache = self.cache.lock();
+        if let Some(entry) = cache.get(image_uri).cloned() {
+            let index = uri_index?.1;
+            match entry {
+                Ok(image) => Ok(ImagePoll::Ready {
+                    image: image.get_image(index),
+                }),
+                Err(err) => Err(LoadError::Loading(err)),
+            }
+        } else {
+            match ctx.try_load_bytes(image_uri) {
+                Ok(BytesPoll::Ready { bytes, .. }) => {
+                    let is_gif = bytes.starts_with(b"GIF87a") || bytes.starts_with(b"GIF89a");
+                    if !is_gif {
+                        return Err(LoadError::NotSupported);
+                    }
+                    let index = uri_index?.1;
```

Comment:
"index" is very generic. Index of what? Something like "frame_index" is more descriptive